File: cpp/LC_23.cpp
#include <vector>
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeKLists(std::vector<ListNode*>& lists) {
        // Time  --- O(2t(lists[0] + ... + lists[t])) -> O(t^2 * avgLen(lists))
        ListNode* head = nullptr;
        int start;
        for (int i = 0; i < lists.size(); i++) {
            if (lists[i] != nullptr) {
                head = lists[i];
                start = i;
                break;
            }
        }
        if (head == nullptr) {
            return nullptr;
        }
        
        for (int i = 0; i < lists.size(); i++) {
            if (lists[i] != nullptr) {
                if (lists[i]->val < head->val) {
                    head = lists[i];
                    start = i;
                }
            }
        }
        lists[start] = lists[start]->next;
        head->next = mergeKLists(lists);
        return head;
    }
};


int main() {
    ListNode n1(1);
    ListNode n2(4);
    ListNode n3(5);
    n2.next = &n3;
    n1.next = &n2;

    ListNode h1(1);
    ListNode h2(3);
    ListNode h3(4);
    h2.next = &h3;
    h1.next = &h2;

    ListNode g1(2);
    ListNode g2(6);
    g1.next = &g2;

    std::vector<ListNode*> arr{&n1, &h1, &g1};

    Solution s;
    ListNode* ans = s.mergeKLists(arr);
    while (ans != nullptr) {
        std::cout << ans->val << "->";
        ans = ans->next;
    }
    std::cout << "NULL\n";
    return 0;
}

File: py/LC_23.py
import heapq


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: list[ListNode]) -> [ListNode]:
        if not lists: return None

        h = []
        heapq.heapify(h)
        for node in lists:
            while node is not None:
                heapq.heappush(h, node.val)
                node = node.next
        if not h: return None

        head = ListNode(heapq.heappop(h))
        cur = head
        while h:
            cur.next = ListNode(heapq.heappop(h))
            cur = cur.next

        return head
        


def main():
    s = Solution()
    
    # [[1,4,5],[1,3,4],[2,6]]
    n1 = ListNode(1,\
                  ListNode(4,\
                           ListNode(5)))
    n2 = ListNode(1,\
                  ListNode(3,\
                           ListNode(4)))
    n3 = ListNode(2, \
                  ListNode(6))
    ans = s.mergeKLists([n1, n2, n3])

    while ans is not None:
        print(f'{ans.val}->', end='')
        ans = ans.next
    print('None')
    
    


if __name__ == "__main__":
    main()